user_util/user_util.py

```diff
@@ -1,3 +1,112 @@
+        and all the historical salts, including the current salt, from oldest to current.
+    """
+    if not isinstance(salt_list, (list, tuple)):
+        raise SALT_LIST_EXCEPTION
+
+    for salt in salt_list:
+        yield retired_username_fmt.format(_compute_retired_hash(username, salt))
+
+
+def get_all_retired_emails(email, salt_list, retired_email_fmt=RETIRED_EMAIL_DEFAULT_FMT):
+    """
+    Returns a generator of possible retired email addresses based on the original email
+    and all the historical salts, from oldest to current.
+    The current salt is assumed to be the last salt in the list.
+
+    Raises :class:`~ValueError` if the salt isn't a list of salts.
+
+    Arguments:
+        email (str): Email address of the user to be retired.
+        salt_list (list/tuple): List of all historical salts.
+
+    Yields:
+        Returns a generator of possible retired email addresses based on the original email
+        and all the historical salts, including the current salt, from oldest to current.
+    """
+    if not isinstance(salt_list, (list, tuple)):
+        raise SALT_LIST_EXCEPTION
+
+    for salt in salt_list:
+        yield retired_email_fmt.format(_compute_retired_hash(email, salt))
+
+
+def get_retired_username(username, salt_list, retired_username_fmt=RETIRED_USERNAME_DEFAULT_FMT):
+    """
+    Returns a retired username based on the original username
+    and all the historical salts, from oldest to current.
+    The current salt is assumed to be the last salt in the list.
+
+    Raises :class:`~ValueError` if the salt isn't a list of salts.
+
+    Arguments:
+        username (str): The name of the user to be retired.
+        salt_list (list/tuple): List of all historical salts.
+
+    Yields:
+        Returns a retired username based on the original username
+        and all the historical salts, including the current salt.
+    """
+    if not isinstance(salt_list, (list, tuple)):
+        raise SALT_LIST_EXCEPTION
+
+    return retired_username_fmt.format(_compute_retired_hash(username, salt_list[-1]))
+
+
+def get_retired_email(email, salt_list, retired_email_fmt=RETIRED_EMAIL_DEFAULT_FMT):
+    """
+    Returns a retired email address based on the original email address
+    and all the historical salts, from oldest to current.
+    The current salt is assumed to be the last salt in the list.
+
+    Raises :class:`~ValueError` if the salt isn't a list of salts.
+
+    Arguments:
+        email (str): Email address of the user to be retired.
+        salt_list (list/tuple): List of all historical salts.
+
+    Yields:
+        Returns a retired email address based on the original email
+        and all the historical salts, including the current salt.
+    """
+    if not isinstance(salt_list, (list, tuple)):
+        raise SALT_LIST_EXCEPTION
+
+    return retired_email_fmt.format(_compute_retired_hash(email, salt_list[-1]))
```

Comment:
My I suggest a function called "is_username_retired" method?

Reply:
@macdiesel The way that method will work involves Django - and I've kept Django out of this module to make it simple as possible. The steps to implement that method:

- Call `get_all_retired_usernames` with the username in question and all secret salts.
- Construct a query against the `User` model, filtering on all the returned retired usernames.
- If one of the retired usernames is found, the username is retired. Else not.

These steps keep the checking to a single DB query and keep a Django dependency out of this module.